src/_P121_HMC5883L.ino

```diff
@@ -0,0 +1,154 @@
+
+#include "_Plugin_Helper.h"
+#ifdef USES_P121
+
+// #######################################################################################################
+// #################### Plugin 121 - Adafruit HMC5883L Driver (3-Axis Magnetometer)               ########
+// #######################################################################################################
+
+/*******************************************************************************
+* Copyright 2021
+* Written by Sven Briesen (adsorstuff@gmail.com)
+* BSD license, all text above must be included in any redistribution
+*
+* Release notes:
+* Adafruit_HMC5883_Unified Library v1.2.0 required (https://github.com/adafruit/Adafruit_HMC5883_Unified)
+* Used P106 BME680 as starting point
+/******************************************************************************/
+
+#include "src/PluginStructs/P121_data_struct.h"
+#include <Adafruit_Sensor.h>
+#include <Adafruit_HMC5883_U.h>
+
+#define PLUGIN_121
+#define PLUGIN_ID_121 121
+#define PLUGIN_NAME_121 "Environment - HCM5883L"
+#define PLUGIN_VALUENAME1_121 "x"
+#define PLUGIN_VALUENAME2_121 "y"
+#define PLUGIN_VALUENAME3_121 "z"
+#define PLUGIN_VALUENAME4_121 "dir"
+
+boolean Plugin_121(uint8_t function, struct EventStruct *event, String &string)
+{
+  boolean success = false;
+
+  switch (function)
+  {
+  case PLUGIN_DEVICE_ADD:
+  {
+    Device[++deviceCount].Number = PLUGIN_ID_121;
+    Device[deviceCount].Type = DEVICE_TYPE_I2C;
+    Device[deviceCount].VType = Sensor_VType::SENSOR_TYPE_QUAD;
+    Device[deviceCount].Ports = 0;
+    Device[deviceCount].PullUpOption = false;
+    Device[deviceCount].InverseLogicOption = false;
+    Device[deviceCount].FormulaOption = true;
+    Device[deviceCount].ValueCount = 4;
+    Device[deviceCount].SendDataOption = true;
+    Device[deviceCount].TimerOption = true;
+    Device[deviceCount].GlobalSyncOption = true;
+    success = true;
+    break;
+  }
+
+  case PLUGIN_GET_DEVICENAME:
+  {
+    string = F(PLUGIN_NAME_121);
+    success = true;
+    break;
+  }
+
+  case PLUGIN_GET_DEVICEVALUENAMES:
+  {
+    strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_121));
+    strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[1], PSTR(PLUGIN_VALUENAME2_121));
+    strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[2], PSTR(PLUGIN_VALUENAME3_121));
+    strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[3], PSTR(PLUGIN_VALUENAME4_121));
+    success = true;
+    break;
+  }
+  case PLUGIN_WEBFORM_SHOW_I2C_PARAMS:
+  {
+    success = true;
+    break;
+  }
+  case PLUGIN_I2C_HAS_ADDRESS:
+  {
+    success = (event->Par1 == 0x1E);
+    break;
+  }
+  case PLUGIN_WEBFORM_LOAD:
+  {
+    addFormFloatNumberBox(F("DeclinationAngle"), F("plugin_121_HMC5883L_decl"), PCONFIG_FLOAT(0), -2.0f,2.0f ,2, 0.01f);
```

Comment:
Wouldn't it be helpful to have  a few more decimals of accuracy then 2 on this field? I'd say the only 2 decimals is rather coarse, so 4 or 5 would fit better, seeing that 1 deg. is only 0.01745 rad. The PCONFIG_FLOAT(n) will happily store those decimals.

Reply:
Yes, sure. I agree.

Reply:
The unit is mentioned to be in degree, so I doubt whether it makes sense to have it in more than 2 decimals.
At 1 meter from the rotation point it only makes a difference of 0.17 mm.
Not sure if a digital sensor can manage this accurately and still be affordable for use cases where ESPEasy will be used :)

Reply:
It needs to be the radians of the degree, which is already 0.0xxx. With 5 digits the user can decide to use 0.04 or 0.04893

Reply:
I can imagine you need the value to be in rads, but don't use rads in the user interface.
So please convert to and from degrees for the user interface.

Reply:
I would leave it and make is more clear, that the user needs to enter radians. Otherwise you will need to convert this value with every cycle.

Reply:
If you convert it to degree when loading the web page (`PLUGIN_WEBFORM_LOAD`) and convert it back when saving it (`PLUGIN_WEBFORM_SAVE`) , then it doesn't need to be converted that often.

Reply:
I need PI for that. Can you tell me, where to look?

Reply:
Found it